collect_app/src/test/java/org/odk/collect/android/provider/FormsProviderTest.java

```diff
@@ -0,0 +1,478 @@
+            assertThat(cursor.getString(cursor.getColumnIndex(JR_VERSION)), is(formVersion));
+            assertThat(cursor.getString(cursor.getColumnIndex(FORM_FILE_PATH)), is(formFile.getName()));
+
+            assertThat(cursor.getString(cursor.getColumnIndex(DATE)), is(notNullValue()));
+            assertThat(cursor.getString(cursor.getColumnIndex(MD5_HASH)), is(md5Hash));
+            assertThat(cursor.getString(cursor.getColumnIndex(JRCACHE_FILE_PATH)), is(md5Hash + ".formdef"));
+            assertThat(cursor.getString(cursor.getColumnIndex(FORM_MEDIA_PATH)), is(mediaPathForFormFile(formFile)));
+        }
+    }
+
+    @Test
+    public void insert_returnsFormUri() {
+        String formId = "external_app_form";
+        String formVersion = "1";
+        String formName = "External app form";
+        File formFile = addFormToFormsDir(formId, formVersion, formName);
+
+        ContentValues values = getContentValues(formId, formVersion, formName, formFile);
+        Uri newFormUri = contentResolver.insert(CONTENT_URI, values);
+
+        try (Cursor cursor = contentResolver.query(newFormUri, null, null, null, null)) {
+            assertThat(cursor.getCount(), is(1));
+        }
+    }
+
+    @Test
+    public void update_updatesForm() {
+        Uri formUri = addFormsToDirAndDb("external_app_form", "1", "External app form");
+
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(LANGUAGE, "English");
+
+        contentResolver.update(formUri, contentValues, null, null);
+        try (Cursor cursor = contentResolver.query(formUri, null, null, null)) {
+            assertThat(cursor.getCount(), is(1));
+
+            cursor.moveToNext();
+            assertThat(cursor.getString(cursor.getColumnIndex(LANGUAGE)), is("English"));
+        }
+    }
+
+    @Test
+    public void update_withSelection_onlyUpdatesMatchingForms() {
+        addFormsToDirAndDb("form1", "1", "Matching form");
+        addFormsToDirAndDb("form2", "1", "Not matching form");
+        addFormsToDirAndDb("form3", "1", "Matching form");
+
+        ContentValues contentValues = new ContentValues();
+        contentValues.put(LANGUAGE, "English");
+
+        contentResolver.update(CONTENT_URI, contentValues, DISPLAY_NAME + "=?", new String[]{"Matching form"});
+        try (Cursor cursor = contentResolver.query(CONTENT_URI, null, null, null)) {
+            assertThat(cursor.getCount(), is(3));
+
+            cursor.moveToNext();
+            if (cursor.getString(cursor.getColumnIndex(DISPLAY_NAME)).equals("Matching form")) {
+                assertThat(cursor.getString(cursor.getColumnIndex(LANGUAGE)), is("English"));
+            } else {
+                assertThat(cursor.isNull(cursor.getColumnIndex(LANGUAGE)), is(true));
+            }
+        }
+    }
+
+    @Test
+    @Ignore("Not sure we need this")
+    public void update_withFormFilePath_deletesFiles_andUpdatesHashAndPaths() {
```

Comment:
Yeah I wrote this test file in a branch off of `v1.30.1` (and everything passed) and then rebased it onto my `content-providers` branch so I definitely had a set of tests for the existing behaviour.

> For this one, I guess it's reasonably kind cleanup but the stakes feel low when it comes to having orphaned form definition files. Am I understanding the implications of not having this behavior correctly?

That's my understanding of it.

> I hadn't thought about this before but when we used to store absolute file paths, an external app could have written files to some random place and had Collect use those, right? Now that we store relative paths, this is no longer possible?

Yeah probably right? Although since the storage migration feature went in, and I'd be surprised if something didn't break given there was then a lot of assumptions around relative paths in the app.